File: scripts/zigux/check-phase14-rollback-threshold-gap.py
#!/usr/bin/env python3
"""PHASE14_CHECK_PACKET=rollback_threshold_gap

Fail-closed checker for the current Phase 14 rollback-threshold automation gap
note. This keeps the dedicated rollback note aligned with the returned shared
smoke route, validator, release-boundary checker, manifest, and current
exact-readback gap split without reopening anchor-local ownership.
"""

from __future__ import annotations

import argparse
import sys
import tempfile
from pathlib import Path

MARKER = "PHASE14_CHECK_PACKET=rollback_threshold_gap"
SMOKE_NOTE_PATH = Path("Documentation/zigux/phase14-end-to-end-smoke-survey.md")
PRODUCTIZATION_GAP_PATH = Path("Documentation/zigux/phase14-productization-gap-survey.md")
SHARED_SMOKE_GAP_PATH = Path("Documentation/zigux/phase14-shared-smoke-current-master-gap.md")
ROLLBACK_CHECKER_PATH = Path(
    "scripts/zigux/check-phase14-rollback-threshold-sequencing.py"
)
VALIDATOR_PATH = Path("scripts/zigux/validate-phase14.py")
RELEASE_BOUNDARY_CHECKER_PATH = Path(
    "scripts/zigux/check-phase14-release-boundary-exact-counts.py"
)
MAKEFILE_PATH = Path("zigux/Makefile")
MANIFEST_PATH = Path("zigux/tests/phase14_end_to_end_smoke_manifest.json")
GAP_NOTE_PATH = Path("Documentation/zigux/phase14-rollback-threshold-automation-gap.md")

ROLLBACK_SMOKE_MARKERS = [
    "- `PHASE14_DIRECT_DOC_PACKET=present`",
    "  * rollback owner: `Repo Tooling Pod`",
    "  * rollback threshold: `0` tolerated same-packet drifts",
    "  * fallback path: keep this shared smoke lane aligned",
    "  * automatic return-to-blocked triggers:",
    "    * rollback-threshold-sequencing drift",
]

DIRECT_PACKET_MARKERS = [
    "`scripts/zigux/check-phase14-rollback-threshold-sequencing.py`",
    "`scripts/zigux/validate-phase14.py`",
    "`scripts/zigux/check-phase14-release-boundary-exact-counts.py`",
    "`zigux/tests/phase14_end_to_end_smoke_manifest.json`",
    "`zigux/Makefile`",
]

EXECUTABLE_GAP_MARKERS = [
    "`zigux/tests/phase14_build.zig`",
    "`zigux/tests/phase14_end_to_end_smoke_survey.zig`",
    "`zigux/tests/phase14_skbuff_bridge.zig`",
    "`zigux/tests/phase14_rcu_tree_survey.zig`",
    "`net/core/skbuff_bridge.zig`",
]

GAP_NOTE_MARKERS = [
    "- `PHASE14_ROLLBACK_THRESHOLD_GAP=present`",
    "- `PHASE14_ROLLBACK_THRESHOLD_GAP_KIND=partial_executable_packet_readback_gap`",
    "- `PHASE14_ROLLBACK_THRESHOLD_GAP_SCOPE=shared_smoke_packet_only`",
    "- `PHASE14_ROLLBACK_THRESHOLD_GAP_STATUS_BUCKET=study_only`",
    "- `PHASE14_ROLLBACK_THRESHOLD_GAP_OWNER=Repo Tooling Pod`",
    "The dedicated rollback story is no longer a checker-local self-test inventory drift.",
    "That means same-lane reminder surfaces should describe rollback-threshold automation as directly readable route, checker, validator, manifest, and Makefile evidence while still keeping the broader executable layer explicit as the remaining readback gap.",
]

PROHIBITED_GAP_NOTE_MARKERS = [
    "makefile_selftest_coverage_drift",
    "`scripts/zigux/check-phase14-tests-readme-smoke-summary.py --self-test`",
    "The live shared packet has advanced, but the dedicated rollback-threshold checker has not yet caught up",
]

MAKEFILE_MARKERS = [
    "phase14-validate:",
    "scripts/zigux/validate-phase14.py --self-test",
    "scripts/zigux/validate-phase14.py",
    "scripts/zigux/check-phase14-rollback-threshold-sequencing.py --self-test",
    "scripts/zigux/check-phase14-rollback-threshold-sequencing.py",
    "scripts/zigux/check-phase14-release-boundary-exact-counts.py --self-test",
    "scripts/zigux/check-phase14-release-boundary-exact-counts.py",
]

VALIDATOR_MARKERS = [
    "PHASE14_VALIDATION=pass",
    "PHASE14_VALIDATOR_SELF_TEST=pass",
    "ROLLBACK_THRESHOLD_SEQUENCING_CHECKER_PATH",
    "RELEASE_BOUNDARY_CHECKER_PATH",
    "END_TO_END_SMOKE_MANIFEST_PATH",
]

RELEASE_BOUNDARY_CHECKER_MARKERS = [
    "PHASE14_CHECK_PACKET=release_boundary_exact_counts",
    "PHASE14_RELEASE_BOUNDARY_EXACT_COUNTS_SELF_TEST=pass",
]

MANIFEST_MARKERS = [
    '"phase14_validate_runs_rollback_threshold_sequencing": true',
    '"phase14_make_smoke_target_present": false',
    '"smoke_commands": [',
    '"make -C zigux phase14-validate"',
]


def read_text(root: Path, rel: Path) -> str:
    return (root / rel).read_text(encoding="utf-8")


def source_text() -> str:
    return Path(__file__).read_text(encoding="utf-8")


def check(root: Path) -> list[str]:
    errors: list[str] = []
    if MARKER not in source_text():
        errors.append("checker marker missing from checker source")

    for rel in [
        SMOKE_NOTE_PATH,
        PRODUCTIZATION_GAP_PATH,
        SHARED_SMOKE_GAP_PATH,
        ROLLBACK_CHECKER_PATH,
        VALIDATOR_PATH,
        RELEASE_BOUNDARY_CHECKER_PATH,
        MAKEFILE_PATH,
        MANIFEST_PATH,
        GAP_NOTE_PATH,
    ]:
        if not (root / rel).exists():
            errors.append(f"missing file: {rel.as_posix()}")
    if errors:
        return errors

    smoke_note = read_text(root, SMOKE_NOTE_PATH)
    productization_gap = read_text(root, PRODUCTIZATION_GAP_PATH)
    shared_smoke_gap = read_text(root, SHARED_SMOKE_GAP_PATH)
    rollback_checker = read_text(root, ROLLBACK_CHECKER_PATH)
    validator = read_text(root, VALIDATOR_PATH)
    release_boundary_checker = read_text(root, RELEASE_BOUNDARY_CHECKER_PATH)
    makefile = read_text(root, MAKEFILE_PATH)
    manifest = read_text(root, MANIFEST_PATH)
    gap_note = read_text(root, GAP_NOTE_PATH)

    for marker in ROLLBACK_SMOKE_MARKERS:
        if marker not in smoke_note:
            errors.append(
                f"missing rollback smoke marker in {SMOKE_NOTE_PATH.as_posix()}: {marker}"
            )

    for marker in DIRECT_PACKET_MARKERS:
        if marker not in gap_note:
            errors.append(f"missing direct packet marker in {GAP_NOTE_PATH.as_posix()}: {marker}")
        if marker not in productization_gap:
            errors.append(
                f"missing productization marker in {PRODUCTIZATION_GAP_PATH.as_posix()}: {marker}"
            )
        if marker not in shared_smoke_gap:
            errors.append(
                f"missing shared-smoke-gap marker in {SHARED_SMOKE_GAP_PATH.as_posix()}: {marker}"
            )

    for marker in EXECUTABLE_GAP_MARKERS:
        if marker not in gap_note:
            errors.append(f"missing executable gap marker in {GAP_NOTE_PATH.as_posix()}: {marker}")
        if marker not in productization_gap:
            errors.append(
                f"missing productization executable-gap marker in {PRODUCTIZATION_GAP_PATH.as_posix()}: {marker}"
            )
        if marker not in shared_smoke_gap:
            errors.append(
                f"missing shared-smoke executable-gap marker in {SHARED_SMOKE_GAP_PATH.as_posix()}: {marker}"
            )

    for marker in GAP_NOTE_MARKERS:
        if marker not in gap_note:
            errors.append(f"missing gap-note marker in {GAP_NOTE_PATH.as_posix()}: {marker}")

    for marker in PROHIBITED_GAP_NOTE_MARKERS:
        if marker in gap_note:
            errors.append(
                f"stale gap-note marker still present in {GAP_NOTE_PATH.as_posix()}: {marker}"
            )

    for marker in MAKEFILE_MARKERS:
        if marker not in makefile:
            errors.append(f"missing Makefile marker in {MAKEFILE_PATH.as_posix()}: {marker}")

    for marker in VALIDATOR_MARKERS:
        if marker not in validator:
            errors.append(f"missing validator marker in {VALIDATOR_PATH.as_posix()}: {marker}")

    for marker in RELEASE_BOUNDARY_CHECKER_MARKERS:
        if marker not in release_boundary_checker:
            errors.append(
                "missing release-boundary checker marker in "
                f"{RELEASE_BOUNDARY_CHECKER_PATH.as_posix()}: {marker}"
            )

    for marker in MANIFEST_MARKERS:
        if marker not in manifest:
            errors.append(f"missing manifest marker in {MANIFEST_PATH.as_posix()}: {marker}")

    if "PHASE14_CHECK_PACKET=rollback_threshold_sequencing" not in rollback_checker:
        errors.append(
            "missing rollback-checker packet marker in "
            f"{ROLLBACK_CHECKER_PATH.as_posix()}"
        )

    return errors


def write(root: Path, rel: Path, text: str) -> None:
    path = root / rel
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(text, encoding="utf-8")


def fixture_smoke_note() -> str:
    return """# Phase 14 End-to-End Smoke Survey

- `PHASE14_DIRECT_DOC_PACKET=present`
  * rollback owner: `Repo Tooling Pod`
  * rollback threshold: `0` tolerated same-packet drifts
  * fallback path: keep this shared smoke lane aligned
  * automatic return-to-blocked triggers:
    * rollback-threshold-sequencing drift
"""


def fixture_productization_gap() -> str:
    return """# Phase 14 Productization Gap Survey

- `scripts/zigux/check-phase14-rollback-threshold-sequencing.py`
- `scripts/zigux/validate-phase14.py`
- `scripts/zigux/check-phase14-release-boundary-exact-counts.py`
- `zigux/tests/phase14_end_to_end_smoke_manifest.json`
- `zigux/Makefile`
- `zigux/tests/phase14_build.zig`
- `zigux/tests/phase14_end_to_end_smoke_survey.zig`
- `zigux/tests/phase14_skbuff_bridge.zig`
- `zigux/tests/phase14_rcu_tree_survey.zig`
- `net/core/skbuff_bridge.zig`
"""


def fixture_shared_smoke_gap() -> str:
    return fixture_productization_gap().replace(
        "# Phase 14 Productization Gap Survey", "# Phase 14 Shared Smoke Current-Master Gap"
    )


def fixture_rollback_checker() -> str:
    return 'PHASE14_CHECK_PACKET=rollback_threshold_sequencing\n'


def fixture_validator() -> str:
    return """PHASE14_VALIDATION=pass
PHASE14_VALIDATOR_SELF_TEST=pass
ROLLBACK_THRESHOLD_SEQUENCING_CHECKER_PATH
RELEASE_BOUNDARY_CHECKER_PATH
END_TO_END_SMOKE_MANIFEST_PATH
"""


def fixture_release_boundary_checker() -> str:
    return """PHASE14_CHECK_PACKET=release_boundary_exact_counts
PHASE14_RELEASE_BOUNDARY_EXACT_COUNTS_SELF_TEST=pass
"""


def fixture_makefile() -> str:
    return """phase14-validate:
\tcd $(ZIGUX_ROOT) && $(PYTHON) scripts/zigux/validate-phase14.py --self-test
\tcd $(ZIGUX_ROOT) && $(PYTHON) scripts/zigux/validate-phase14.py
\tcd $(ZIGUX_ROOT) && $(PYTHON) scripts/zigux/check-phase14-rollback-threshold-sequencing.py --self-test
\tcd $(ZIGUX_ROOT) && $(PYTHON) scripts/zigux/check-phase14-rollback-threshold-sequencing.py
\tcd $(ZIGUX_ROOT) && $(PYTHON) scripts/zigux/check-phase14-release-boundary-exact-counts.py --self-test
\tcd $(ZIGUX_ROOT) && $(PYTHON) scripts/zigux/check-phase14-release-boundary-exact-counts.py
"""


def fixture_manifest() -> str:
    return """{
  "smoke_commands": [
    "make -C zigux phase14-validate"
  ],
  "survey_summary": {
    "phase14_make_smoke_target_present": false,
    "phase14_validate_runs_rollback_threshold_sequencing": true
  }
}
"""


def fixture_gap_note() -> str:
    return """# Phase 14 Rollback-Threshold Automation Gap

## Status

- `PHASE14_ROLLBACK_THRESHOLD_GAP=present`
- `PHASE14_ROLLBACK_THRESHOLD_GAP_KIND=partial_executable_packet_readback_gap`
- `PHASE14_ROLLBACK_THRESHOLD_GAP_SCOPE=shared_smoke_packet_only`
- `PHASE14_ROLLBACK_THRESHOLD_GAP_STATUS_BUCKET=study_only`
- `PHASE14_ROLLBACK_THRESHOLD_GAP_OWNER=Repo Tooling Pod`

The dedicated rollback story is no longer a checker-local self-test inventory drift.

- `scripts/zigux/check-phase14-rollback-threshold-sequencing.py`
- `scripts/zigux/validate-phase14.py`
- `scripts/zigux/check-phase14-release-boundary-exact-counts.py`
- `zigux/tests/phase14_end_to_end_smoke_manifest.json`
- `zigux/Makefile`
- `zigux/tests/phase14_build.zig`
- `zigux/tests/phase14_end_to_end_smoke_survey.zig`
- `zigux/tests/phase14_skbuff_bridge.zig`
- `zigux/tests/phase14_rcu_tree_survey.zig`
- `net/core/skbuff_bridge.zig`

That means same-lane reminder surfaces should describe rollback-threshold automation as directly readable route, checker, validator, manifest, and Makefile evidence while still keeping the broader executable layer explicit as the remaining readback gap.
"""


def run_self_test() -> int:
    cases = 5
    with tempfile.TemporaryDirectory() as tmp:
        root = Path(tmp)
        write(root, SMOKE_NOTE_PATH, fixture_smoke_note())
        write(root, PRODUCTIZATION_GAP_PATH, fixture_productization_gap())
        write(root, SHARED_SMOKE_GAP_PATH, fixture_shared_smoke_gap())
        write(root, ROLLBACK_CHECKER_PATH, fixture_rollback_checker())
        write(root, VALIDATOR_PATH, fixture_validator())
        write(root, RELEASE_BOUNDARY_CHECKER_PATH, fixture_release_boundary_checker())
        write(root, MAKEFILE_PATH, fixture_makefile())
        write(root, MANIFEST_PATH, fixture_manifest())
        write(root, GAP_NOTE_PATH, fixture_gap_note())
        errors = check(root)
        if errors:
            print("PHASE14_ROLLBACK_THRESHOLD_GAP_SELF_TEST=fail")
            for error in errors:
                print(error)
            return 1

        write(root, GAP_NOTE_PATH, fixture_gap_note().replace(
            "- `PHASE14_ROLLBACK_THRESHOLD_GAP_OWNER=Repo Tooling Pod`\n", "", 1
        ))
        if not check(root):
            print("PHASE14_ROLLBACK_THRESHOLD_GAP_SELF_TEST=fail")
            print("expected missing gap-note owner failure")
            return 1

        write(root, GAP_NOTE_PATH, fixture_gap_note())
        write(root, MANIFEST_PATH, fixture_manifest().replace(
            '"phase14_validate_runs_rollback_threshold_sequencing": true', ""
        ))
        if not check(root):
            print("PHASE14_ROLLBACK_THRESHOLD_GAP_SELF_TEST=fail")
            print("expected missing manifest sequencing marker failure")
            return 1

        write(root, MANIFEST_PATH, fixture_manifest())
        write(root, GAP_NOTE_PATH, fixture_gap_note().replace(
            "partial_executable_packet_readback_gap", "makefile_selftest_coverage_drift", 1
        ))
        if not check(root):
            print("PHASE14_ROLLBACK_THRESHOLD_GAP_SELF_TEST=fail")
            print("expected stale gap-kind failure")
            return 1

        write(root, GAP_NOTE_PATH, fixture_gap_note())
        write(root, MAKEFILE_PATH, fixture_makefile().replace(
            "\tcd $(ZIGUX_ROOT) && $(PYTHON) scripts/zigux/check-phase14-rollback-threshold-sequencing.py --self-test\n",
            "",
            1,
        ))
        if not check(root):
            print("PHASE14_ROLLBACK_THRESHOLD_GAP_SELF_TEST=fail")
            print("expected missing Makefile rollback self-test failure")
            return 1

        write(root, MAKEFILE_PATH, fixture_makefile())
        write(root, PRODUCTIZATION_GAP_PATH, fixture_productization_gap().replace(
            "- `zigux/tests/phase14_end_to_end_smoke_survey.zig`\n", "", 1
        ))
        if not check(root):
            print("PHASE14_ROLLBACK_THRESHOLD_GAP_SELF_TEST=fail")
            print("expected missing productization executable-gap marker failure")
            return 1

    print("PHASE14_ROLLBACK_THRESHOLD_GAP_SELF_TEST=pass")
    print(f"PHASE14_ROLLBACK_THRESHOLD_GAP_SELF_TEST_CASES={cases}")
    return 0


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--self-test", action="store_true")
    return parser.parse_args()


def main() -> int:
    args = parse_args()
    if args.self_test:
        return run_self_test()

    errors = check(Path.cwd())
    if errors:
        for error in errors:
            print(error, file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/zigux/check_phase14_rollback_threshold_gap.zig
const std = @import("std");
const Io = std.Io;
const guard = @import("zigux_guard.zig");

pub const pass_marker = "PHASE14_ROLLBACK_THRESHOLD_GAP_SELF_TEST=pass";

const ROLLBACK_NOTE_MARKERS = [_][]const u8{
    "- rollback owner: `Repo Tooling Pod`",
    "- rollback threshold: `0` tolerated same-packet drifts across anchor-local manifests, anchor-local survey notes, the compile shard matrix, and shared replay wiring",
    "- fallback path: keep this shared smoke lane parked and rerun `make -C zigux phase14-validate` before reopening any anchor-local or shared follow-up",
    "- automatic return-to-blocked triggers:",
    "anchor-local manifest drift",
    "anchor-local survey note drift",
    "compile shard matrix drift",
    "shared replay wiring drift",
};

const MISSING_EXECUTABLE_MARKERS = [_][]const u8{
    "`scripts\zigux/validate_phase14.zig`",
    "`scripts/zigux/check_phase14_release_boundary_exact_counts.zig`",
    "`zigux/tests/phase14_build.zig`",
    "`zigux/tests/phase14_end_to_end_smoke_manifest.json`",
    "`zigux/tests/phase14_end_to_end_smoke_survey.zig`",
};

const GAP_NOTE_MARKERS = [_][]const u8{
    "- `PHASE14_ROLLBACK_THRESHOLD_GAP=present`",
    "- `PHASE14_ROLLBACK_THRESHOLD_GAP_KIND=executable_packet_readback_gap`",
    "- `PHASE14_ROLLBACK_THRESHOLD_GAP_SCOPE=shared_smoke_packet_only`",
    "- `PHASE14_ROLLBACK_THRESHOLD_GAP_STATUS_BUCKET=study_only`",
    "- `PHASE14_ROLLBACK_THRESHOLD_GAP_OWNER=Repo Tooling Pod`",
    "The directly readable rollback-threshold packet is stronger than an older docs-absence claim.",
    "But the executable rollback-threshold packet members still return missing-path results on the same exact contents path:",
    "The remaining same-lane gap is no longer a smaller Makefile-self-test inventory mismatch inside `scripts/zigux/check_phase14_rollback_threshold_sequencing.zig`.",
    "tighten broader Phase 14 reminder surfaces so they name the rollback-threshold note/checker layer as directly readable while keeping the executable layer explicit as the remaining gap.",
};

const PROHIBITED_GAP_NOTE_MARKERS = [_][]const u8{
    "makefile_selftest_coverage_drift",
    "Refresh `scripts/zigux/check_phase14_rollback_threshold_sequencing.zig`",
    "`scripts/zigux/check_phase14_tests_readme_smoke_summary.zig --self-test`",
};

const ROLLBACK_CHECKER_MARKERS = [_][]const u8{
    "ROLLBACK_OWNER = \"Repo Tooling Pod\"",
    "ROLLBACK_THRESHOLD_MARKER =",
    "ROLLBACK_FALLBACK_PATH_MARKER =",
    "ROLLBACK_TRIGGER_MARKERS = [",
    "\"  - anchor-local manifest drift\"",
    "\"  - anchor-local survey note drift\"",
    "\"  - compile shard matrix drift\"",
    "\"  - shared replay wiring drift\"",
};

const ROLLBACK_TRIGGER_MARKERS = [_][]const u8{
    ",\n    '-anchor-localmanifestdrift',\n    '-anchor-localsurveynotedrift',\n    '-compileshardmatrixdrift',\n    '-sharedreplaywiringdrift',\n]\n\n\ndef read_text(root: Path, rel: Path) -> str:\n    return (root / rel).read_text(encoding=utf-8)\n\n\ndef source_text() -> str:\n    return Path(__file__).read_text(encoding=utf-8)\n\n\ndef check(root: Path) -> list[str]:\n    errors: list[str] = []\n    if MARKER not in source_text():\n        errors.append(checkermarkermissingfromcheckersource)\n\n    for rel in [\n        SMOKE_NOTE_PATH,\n        PRODUCTIZATION_GAP_PATH,\n        SHARED_SMOKE_GAP_PATH,\n        ROLLBACK_CHECKER_PATH,\n        GAP_NOTE_PATH,\n    ]:\n        if not (root / rel).exists():\n            errors.append(fmissingfile:{rel.as_posix()})\n    if errors:\n        return errors\n\n    smoke_note = read_text(root, SMOKE_NOTE_PATH)\n    productization_gap = read_text(root, PRODUCTIZATION_GAP_PATH)\n    shared_smoke_gap = read_text(root, SHARED_SMOKE_GAP_PATH)\n    rollback_checker = read_text(root, ROLLBACK_CHECKER_PATH)\n    gap_note = read_text(root, GAP_NOTE_PATH)\n\n    for marker in ROLLBACK_NOTE_MARKERS:\n        if marker not in smoke_note:\n            errors.append(\n                fmissingrollback-notemarkerin{SMOKE_NOTE_PATH.as_posix()}:{marker}\n            )\n\n    for marker in MISSING_EXECUTABLE_MARKERS:\n        if marker not in productization_gap:\n            errors.append(\n                fmissingproductization-gapmarkerin{PRODUCTIZATION_GAP_PATH.as_posix()}:{marker}\n            )\n        if marker not in shared_smoke_gap:\n            errors.append(\n                fmissingshared-smoke-gapmarkerin{SHARED_SMOKE_GAP_PATH.as_posix()}:{marker}\n            )\n        if marker not in gap_note:\n            errors.append(\n                fmissingrollback-gapnotemarkerin{GAP_NOTE_PATH.as_posix()}:{marker}\n            )\n\n    for marker in GAP_NOTE_MARKERS:\n        if marker not in gap_note:\n            errors.append(fmissinggap-notemarkerin{GAP_NOTE_PATH.as_posix()}:{marker})\n\n    for marker in PROHIBITED_GAP_NOTE_MARKERS:\n        if marker in gap_note:\n            errors.append(\n                fstalegap-notemarkerstillpresentin{GAP_NOTE_PATH.as_posix()}:{marker}\n            )\n\n    for marker in ROLLBACK_CHECKER_MARKERS:\n        if marker not in rollback_checker:\n            errors.append(\n                missingrollback-checkermarkerin\n                f{ROLLBACK_CHECKER_PATH.as_posix()}:{marker}\n            )\n\n    return errors\n\n\ndef write(root: Path, rel: Path, text: str) -> None:\n    path = root / rel\n    path.parent.mkdir(parents=True, exist_ok=True)\n    path.write_text(text, encoding=utf-8)\n\n\ndef fixture_smoke_note() -> str:\n    return #Phase14End-to-EndSmokeSurvey-rollbackowner:`RepoToolingPod`-rollbackthreshold:`0`toleratedsame-packetdriftsacrossanchor-localmanifests",
    "anchor-localsurveynotes",
    "thecompileshardmatrix",
    "andsharedreplaywiring-fallbackpath:keepthissharedsmokelaneparkedandrerun`make-Cziguxphase14-validate`beforereopeninganyanchor-localorsharedfollow-up-automaticreturn-to-blockedtriggers:-anchor-localmanifestdrift-anchor-localsurveynotedrift-compileshardmatrixdrift-sharedreplaywiringdrift\n\n\ndef fixture_productization_gap() -> str:\n    return #Phase14ProductizationGapSurvey-`scripts\zigux/validate_phase14.zig`-`scripts/zigux/check_phase14_release_boundary_exact_counts.zig`-`zigux/tests/phase14_build.zig`-`zigux/tests/phase14_end_to_end_smoke_manifest.json`-`zigux/tests/phase14_end_to_end_smoke_survey.zig`\n\n\ndef fixture_shared_smoke_gap() -> str:\n    return #Phase14SharedSmokeCurrent-MasterGap-`scripts\zigux/validate_phase14.zig`-`scripts/zigux/check_phase14_release_boundary_exact_counts.zig`-`zigux/tests/phase14_build.zig`-`zigux/tests/phase14_end_to_end_smoke_manifest.json`-`zigux/tests/phase14_end_to_end_smoke_survey.zig`\n\n\ndef fixture_rollback_checker() -> str:\n    return #!/usr/bin/envpython3ROLLBACK_OWNER=\\Repo Tooling Pod\"\nROLLBACK_THRESHOLD_MARKER = \"threshold\"\nROLLBACK_FALLBACK_PATH_MARKER = \"fallback\"\nROLLBACK_TRIGGER_MARKERS = [\n    \"  - anchor-local manifest drift\",\n    \"  - anchor-local survey note drift\",\n    \"  - compile shard matrix drift\",\n    \"  - shared replay wiring drift\",\n]\ndeffixture_gap_note()->str:return# Phase 14 Rollback-Threshold Automation Gap\n\n## Status\n\n- `PHASE14_ROLLBACK_THRESHOLD_GAP=present`\n- `PHASE14_ROLLBACK_THRESHOLD_GAP_KIND=executable_packet_readback_gap`\n- `PHASE14_ROLLBACK_THRESHOLD_GAP_SCOPE=shared_smoke_packet_only`\n- `PHASE14_ROLLBACK_THRESHOLD_GAP_STATUS_BUCKET=study_only`\n- `PHASE14_ROLLBACK_THRESHOLD_GAP_OWNER=Repo Tooling Pod`\n\n## Why this gap note exists\n\nThe directly readable rollback-threshold packet is stronger than an older docs-absence claim.\nBut the executable rollback-threshold packet members still return missing-path results on the same exact contents path:\n\n- `scripts\zigux/validate_phase14.zig`\n- `scripts/zigux/check_phase14_release_boundary_exact_counts.zig`\n- `zigux/tests/phase14_build.zig`\n- `zigux/tests/phase14_end_to_end_smoke_manifest.json`\n- `zigux/tests/phase14_end_to_end_smoke_survey.zig`\n\n## Current bounded gap\n\nThe remaining same-lane gap is no longer a smaller Makefile-self-test inventory mismatch inside `scripts/zigux/check_phase14_rollback_threshold_sequencing.zig`.\n\n## Next bounded fix\n\nEither re-materialize the missing executable packet members above on current `master`, or tighten broader Phase 14 reminder surfaces so they name the rollback-threshold note/checker layer as directly readable while keeping the executable layer explicit as the remaining gap.\ndefrun_self_test()->int:cases=5withtempfile.TemporaryDirectory()astmp:root=Path(tmp)write(root",
    "SMOKE_NOTE_PATH",
    "fixture_smoke_note())write(root",
    "PRODUCTIZATION_GAP_PATH",
    "fixture_productization_gap())write(root",
    "SHARED_SMOKE_GAP_PATH",
    "fixture_shared_smoke_gap())write(root",
    "ROLLBACK_CHECKER_PATH",
    "fixture_rollback_checker())write(root",
    "GAP_NOTE_PATH",
    "fixture_gap_note())errors=check(root)iferrors:print(PHASE14_ROLLBACK_THRESHOLD_GAP_SELF_TEST=fail)forerrorinerrors:print(error)return1write(root",
    "PRODUCTIZATION_GAP_PATH",
    "fixture_productization_gap().replace(- `zigux/tests/phase14_end_to_end_smoke_manifest.json`n",
    "1)",
    ")ifnotcheck(root):print(PHASE14_ROLLBACK_THRESHOLD_GAP_SELF_TEST=fail)print(expected missing executable-marker failure)return1write(root",
    "PRODUCTIZATION_GAP_PATH",
    "fixture_productization_gap())write(root",
    "GAP_NOTE_PATH",
    "fixture_gap_note()+nRefresh `scripts/zigux/check_phase14_rollback_threshold_sequencing.zig`n",
    ")ifnotcheck(root):print(PHASE14_ROLLBACK_THRESHOLD_GAP_SELF_TEST=fail)print(expected stale gap-note guidance failure)return1write(root",
    "GAP_NOTE_PATH",
    "fixture_gap_note())write(root",
    "SMOKE_NOTE_PATH",
    "fixture_smoke_note().replace(- automatic return-to-blocked triggers:n",
    "1)",
    ")ifnotcheck(root):print(PHASE14_ROLLBACK_THRESHOLD_GAP_SELF_TEST=fail)print(expected missing rollback-trigger heading failure)return1write(root",
    "SMOKE_NOTE_PATH",
    "fixture_smoke_note())write(root",
    "ROLLBACK_CHECKER_PATH",
    "fixture_rollback_checker().replace(    \"  - shared replay wiring drift\",n",
    "1)",
    ")ifnotcheck(root):print(PHASE14_ROLLBACK_THRESHOLD_GAP_SELF_TEST=fail)print(expected missing rollback-checker trigger failure)return1print(PHASE14_ROLLBACK_THRESHOLD_GAP_SELF_TEST=pass)print(fPHASE14_ROLLBACK_THRESHOLD_GAP_SELF_TEST_CASES={cases})return0defparse_args()->argparse.Namespace:parser=argparse.ArgumentParser()parser.add_argument(--self-test",
    "action=store_true)returnparser.parse_args()defmain()->int:args=parse_args()ifargs.self_test:returnrun_self_test()errors=check(Path.cwd())iferrors:forerrorinerrors:print(error",
};

const MARKER = [_][]const u8{
    "PHASE14_CHECK_PACKET=rollback_threshold_gap",
};

pub fn checkText(text: []const u8) guard.GuardError!void {
    for (ROLLBACK_NOTE_MARKERS) |marker| try guard.requireMarker(text, marker);
    for (MISSING_EXECUTABLE_MARKERS) |marker| try guard.requireMarker(text, marker);
    for (GAP_NOTE_MARKERS) |marker| try guard.requireMarker(text, marker);
    for (PROHIBITED_GAP_NOTE_MARKERS) |marker| try guard.requireMarker(text, marker);
    for (ROLLBACK_CHECKER_MARKERS) |marker| try guard.requireMarker(text, marker);
    for (ROLLBACK_TRIGGER_MARKERS) |marker| try guard.requireMarker(text, marker);
    for (MARKER) |marker| try guard.requireMarker(text, marker);
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();
    const io = std.Io.Threaded.init(allocator, .{});
    defer io.deinit();
    const args = try std.process.argsAlloc(allocator);
    defer std.process.argsFree(allocator, args);

    var self_test = false;
    for (args[1..]) |arg| {
        if (std.mem.eql(u8, arg, "--self-test")) self_test = true;
    }

    if (self_test) {
        try checkText("");
        try guard.printLine(io, "{s}", .{pass_marker});
        return;
    }

    const root = try guard.repoRootFromScript(allocator);
    defer allocator.free(root);
    const workflow_rel = ".github/workflows/zigux-bootstrap.yml";
    const workflow_path = try std.fmt.allocPrint(allocator, "{s}/{s}", .{ root, workflow_rel });
    defer allocator.free(workflow_path);
    const text = try guard.readUtf8File(io, allocator, workflow_path);
    defer allocator.free(text);
    try checkText(text);
    try guard.printLine(io, "{s}", .{pass_marker});
}
